src/lib.rs

```diff
@@ -14,6 +14,166 @@
+                    .flatten()
+                    .map(|pub_key| (idx, beefy_mmr::BeefyEcdsaToEthereum::convert(pub_key)))
+                    .ok_or_else(|| BeefyClientError::InvalidSignature)
+            })
+            .collect::<Result<Vec<_>, BeefyClientError>>()?;
+
+        let mut authorities_changed = false;
+
+        let authorities_merkle_proof =
+            MerkleProof::<KeccakHasher>::new(mmr_update.authority_proof.clone());
+        let authority_leaf_indices = authority_addresses_and_indices
+            .iter()
+            .cloned()
+            .map(|x| x.0 as usize)
+            .collect::<Vec<_>>();
+        let authority_leaves = authority_addresses_and_indices
+            .into_iter()
+            .map(|x| keccak_256(&x.1))
+            .collect::<Vec<_>>();
+
+        // Verify mmr_update.authority_proof against store root hash
+        if current_authority_set.id == validator_set_id {
+            let root_hash = current_authority_set.root;
+            if !authorities_merkle_proof.verify(
+                root_hash.into(),
+                &authority_leaf_indices,
+                &authority_leaves,
+                current_authority_set.len as usize,
+            ) {
+                return Err(BeefyClientError::InvalidAuthorityProof);
+            }
+        } else if next_authority_set.id == validator_set_id {
+            let root_hash = next_authority_set.root;
+            if !authorities_merkle_proof.verify(
+                root_hash.into(),
+                &authority_leaf_indices,
+                &authority_leaves,
+                next_authority_set.len as usize,
+            ) {
+                return Err(BeefyClientError::InvalidAuthorityProof);
+            }
+            authorities_changed = true;
+        }
+
+        let latest_beefy_height = <Store as StorageRead>::latest_height()?;
+
+        if mmr_update.signed_commitment.commitment.block_number <= latest_beefy_height {
+            return Err(BeefyClientError::InvalidMmrUpdate);
+        }
+
+        // Move on to verify mmr_proof
+
+        let proof = pallet_mmr_primitives::Proof {
+            leaf_index: mmr_update.latest_mmr_leaf_with_index.index,
+            // we treat this leaf as the latest leaf in the mmr
+            leaf_count: mmr_update.latest_mmr_leaf_with_index.index + 1,
+            items: mmr_update.mmr_proof.clone(),
+        };
+
+        let node = pallet_mmr_primitives::DataOrHash::Data(
+            mmr_update.latest_mmr_leaf_with_index.leaf.encode(),
+        );
+        pallet_mmr::verify_leaf_proof::<sp_runtime::traits::Keccak256, _>(
+            mmr_root_hash.into(),
+            node,
+            proof,
+        )
+        .map_err(|_| BeefyClientError::InvalidMmrProof)?;
+
+        <Store as StorageWrite>::set_latest_height(
+            mmr_update.signed_commitment.commitment.block_number,
+        )?;
+        <Store as StorageWrite>::set_latest_mmr_root_hash(mmr_root_hash.into())?;
```

Comment:
we definitely want these two calls to happen as a transaction, so maybe batch them into 1 struct with a single method.